tests/test_download_images.py

```diff
@@ -0,0 +1,284 @@
+import unittest
+from unittest.mock import patch, MagicMock, mock_open, call
+import pandas as pd
+import os
+import io
+import shutil
+from io import BytesIO
+from PIL import Image
+import requests
+import base64
+from cautiousrobot.__main__ import download_images, main
+
+class TestDownload(unittest.TestCase):
+    def setUp(self):
+        self.DUMMY_DATA = pd.DataFrame(data={
+            "filename": ["test_file1", "test_file2"],
+            "file_url": ["http://test_url1.com/image.jpg", "http://test_url2.com/image.jpg"],
+            "subfolder": ["test_subfolder1", "test_subfolder2"]
+        })
+        self.IMG_DIR = "test_dir"
+        self.LOG_FILEPATH = "test_log_path.jsonl"
+        self.ERROR_LOG_FILEPATH = "test_error_log_path.jsonl"
+        self.DOWNSAMPLE_PATH = "test_downsample_dir"
+        self.DOWNSAMPLE_SIZE = 100
+
+        os.makedirs(self.IMG_DIR, exist_ok=True)
```

Comment:
It can't find the directories for `tests/test_download_images.py`. I'm guessing it's the timing of creation and tear down or that isn't actually happening?

Reply:
Thank you! I am guessing it might because the 'downsample_dir' was used before it is defined in the code. I will investigate further.

Reply:
Hello @egrace479 The error 'UnboundLocalError for downsample_dir' suggests that downsample_dir is being referenced before it has been assigned a value. Could you please review this? It might be a defect that needs to be addressed. Thank you! 

There are other failed test cases with retry and logging, I will investigate further.

Reply:
Thank you, yes I found the error with the downsampling directory. I set `downsample_dir_path` in the same manner as `image_dir_path` at the beginning of `download_images` (so it's always `downsample_dir_path`, then gets the subfolder addition if subfolders are selected).